pkg/install/cleanup_unix.go

```diff
@@ -0,0 +1,209 @@
+}
+
+func (c *CleanUpConfig) cleanupNetworkNamespace() error {
+	var msg []string
+
+	mounter := mount.New("")
+	procMounts, err := mounter.List()
+	if err != nil {
+		return err
+	}
+	// search and unmount namespace mounts
+	for _, v := range procMounts {
+		if strings.Contains(v.Path, "run/netns") {
+			if err = mounter.Unmount(v.Path); err != nil {
+				msg = append(msg, err.Error())
+			}
+			if err := os.RemoveAll(v.Path); err != nil {
+				msg = append(msg, err.Error())
+			}
+		}
+	}
+	if len(msg) > 0 {
+		return fmt.Errorf("%v", strings.Join(msg, ", "))
+	}
+	return nil
+}
+
+func (c *CleanUpConfig) stopAllContainers() error {
+	var msg []string
+
+	containers, err := c.listContainers()
+	if err != nil {
+		return err
+	}
+
+	for _, container := range containers {
+		logrus.Debugf("stopping container: %v", container)
+		out, err := exec.Command(c.crictlBinPath, "-r", c.criSocketPath, "stopp", container).CombinedOutput()
+		if err != nil {
+			if strings.Contains(string(out), "443: connect: connection refused") {
+				// on a single node instance, we will see "connection refused" error. this is to be expected
+				// since we're deleting the API pod itself. so we're ignoring this error
+				logrus.Debugf("ignoring container stop err: %v", string(out))
+			} else {
+				fmtError := fmt.Errorf("failed to stop running pod %v: output: %v, err: %v", container, string(out), err)
+				msg = append(msg, fmtError.Error())
+			}
+		}
+	}
+	if len(msg) > 0 {
+		return fmt.Errorf("%v", strings.Join(msg, ", "))
+	}
+	return nil
+}
+
+func (c *CleanUpConfig) removeAllContainers() error {
+	var msg []string
+
+	containers, err := c.listContainers()
+	if err != nil {
+		return err
+	}
+
+	for _, container := range containers {
+		out, err := exec.Command(c.crictlBinPath, "-r", c.criSocketPath, "rmp", container).CombinedOutput()
+		if err != nil {
+			fmtError := fmt.Errorf("failed to stop running pod %v: output: %v, err: %v", container, string(out), err)
+			msg = append(msg, fmtError.Error())
+		}
+	}
+	if len(msg) > 0 {
+		return fmt.Errorf("%v", strings.Join(msg, ", "))
+	}
+	return nil
+}
+
+func (c *CleanUpConfig) startContainerd() error {
+	args := []string{
+		fmt.Sprintf("--root=%s", filepath.Join(c.dataDir, "containerd")),
+		fmt.Sprintf("--state=%s", filepath.Join(c.runDir, "containerd")),
+		fmt.Sprintf("--address=%s", c.containerdSockerPath),
+		"--config=/etc/k0s/containerd.toml",
+	}
+	cmd := exec.Command(c.containerdBinPath, args...)
+	if err := cmd.Start(); err != nil {
+		return fmt.Errorf("failed to start containerd: %v", err)
+	}
+	go func() {
+		for {
+			select {
+			case <-c.quit:
+				logrus.Debug("stopping clean-up instance of containerd...")
+				if err := cmd.Process.Kill(); err != nil {
+					logrus.Errorf("failed to kill containerd: %v", err)
+				}
+			default:
+				continue
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (c *CleanUpConfig) stopContainerd() {
+	logrus.Debug("attempting to stop containerd")
+	c.quit <- true
```

Comment:
I think we technically should wait for the containerd process to exit and collect the exit status, or we'll have a containerd zombie process. I don't know if it matters though, since the uninstall is a short lived process.

Reply:
The function now uses os.SIGINT (https://golang.org/src/os/exec.go?s=2142:2233#L59) to stop containerd. If the process is still up, it will wait 3 seconds and will send SIGKILL.